src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,506 @@
+    private int kvStoreInsertionPos;
+    // This tracks the number of occupied slots at which we should rebuild the index array
+    // For example, at a load factor of 0.75 and an index of total size 100 slot, that number is 75 slots
+    private int numSlotsForIndexRebuild;
+
+
+    // Each slot in the index array can be in one of 3 states, depending on the value of its second (offset) field :
+    // Offset >= 1 : Filled, the data in the hash field and the index field is valid. Subtracting one from the offset
+    // will yield a valid index into the KV array.
+    // Offset == 0 : Unused, the data in the hash field and the index field is NOT valid, and the slot was never filled
+    // with valid data.
+    // Offset == -1 : Deleted, the data in the hash field and the index field is NOT valid, but the slot used to be
+    // occupied with valid data before.
+    private static final int INDEX_SLOT_UNUSED_MARKER = 0;
+    private static final int INDEX_SLOT_DELETED_MARKER = -1;
+
+    // Number of hash entries, not array positions (in general, capacities and sizes are always in entries)
+    public static final int DEFAULT_INITIAL_CAPACITY = 8;
+    public static final float THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD = 0.7f;
+
+    // temporary debugging aid so that its obvious the class is being used without being spam
+    private static boolean LOG_FIRST_ACCESS = false;
+
+    public CompactHashStore() {
+        this(DEFAULT_INITIAL_CAPACITY);
+    }
+
+    // private non-allocating constructor for .copy(), not part of the interface
+    private CompactHashStore(int[] index, Object[] kvs, int kvStoreInsertionPos, int numSlotsForIndexRebuild) {
+        this.index = index;
+        this.kvStore = kvs;
+        this.kvStoreInsertionPos = kvStoreInsertionPos;
+        this.numSlotsForIndexRebuild = numSlotsForIndexRebuild;
+    }
+
+    public CompactHashStore(int capacity) {
+        // temporary and ugly
+        if (LOG_FIRST_ACCESS) {
+            System.out.println("~~~~~~~~~~~~~~~~~~~~~~~ Using Compact Hashes from Hash Tables That Don't Hate You.");
+            LOG_FIRST_ACCESS = false; // without this we would spam the output stream
+        }
+        assertConstructorPreconditions(capacity);
+
+        int kvCapacity = roundUpwardsToNearestPowerOf2(capacity);
+        // the index array needs to be a little bit sparse for good performance (i.e. low load factor)
+        // so to store 8 key-value entries an index of exactly 8 slots is too crowded
+        int indexCapacity = 2 * kvCapacity;
+
+        // All 0s by default, so all slots are marked empty for free
+        this.index = new int[2 * indexCapacity];
+        this.kvStore = new Object[2 * kvCapacity];
+        this.kvStoreInsertionPos = 0; // always increases by 2, never decreases
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
```

Comment:
I don't know if you plan to adjust the formula, but since `this.index.length` is `2 * indexCapacity` and then to calculate this value you divide by 2, would it make more sense to swap it with `indexCapacity * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD`?